vicodyn/src/vicodyn/proxy.cpp

```diff
@@ -108,26 +108,105 @@ class safe_stream_t {
     }
 };
 
-class vicodyn_dispatch_t : public std::enable_shared_from_this<vicodyn_dispatch_t> {
-    using protocol = io::protocol<app_tag>::scope;
-    using clock_t = peers_t::clock_t;
+class buffer_t {
+    std::string event_;
+    hpack::headers_t headers_;
 
-    struct error_buffer_t {
-        hpack::headers_t headers;
-        std::error_code ec;
-        std::string msg;
-    };
+    std::vector<std::pair<hpack::headers_t, std::string>> chunks_;
+    boost::optional<hpack::headers_t> choke_;
+    bool failed_ = false;
+    bool overfull_ = false;
 
-    struct buffer_t {
-        std::string event;
-        hpack::headers_t headers;
+    std::size_t capacity_;
 
-        std::vector<std::string> chunks;
-        std::vector<hpack::headers_t> chunk_headers;
 
-        boost::optional<error_buffer_t> error;
-        boost::optional<hpack::headers_t> choke;
-    };
+private:
+    auto try_account(const hpack::headers_t& headers, std::size_t data_size = 0) -> bool {
+        if (overfull()) {
+            return false;
+        }
+        for (const auto& header : headers) {
+            data_size += header.http2_size();
+        }
+        if (data_size > capacity_) {
+            overfull_ = true;
+            clear();
+            return false;
+        }
+        capacity_ -= data_size;
+        return true;
+    }
+
+
+public:
+    buffer_t(std::size_t capacity)
+        : capacity_(capacity) {
+    }
+
+    auto overfull() const -> bool {
+        return overfull_;
+    }
+
+    auto failed() const -> bool {
+        return failed_;
+    }
+
+    auto store_event(const hpack::headers_t& headers, const std::string& event) -> void {
+        if (try_account(headers, event.size())) {
+            headers_ = headers;
+            event_ = event;
+        }
+    }
+
+    auto event() const -> const std::string& {
+        return event_;
+    }
+
+    auto headers() const -> const hpack::headers_t& {
+        return headers_;
+    }
+
+    auto chunk(const hpack::headers_t& headers, const std::string& data) -> void {
+        if (try_account(headers, data.size())) {
+            chunks_.emplace_back(headers, data);
+        }
+    }
+
+    auto close(const hpack::headers_t& headers) -> void {
+        if (try_account(headers)) {
+            choke_ = headers;
+        }
+    }
+
+    auto error(const hpack::headers_t& /*headers*/, const std::error_code& /*ec*/, const std::string& /*msg*/) -> void {
```

Comment:
Why isn't it just `auto error() -> void`?

Reply:
For similarity of interfaces with other classes. Ok, I'll fix it.